indy_node/server/request_handlers/domain_req_handlers/context_handler.py

```diff
@@ -0,0 +1,134 @@
+        validate_meta(meta)
+        data = request.operation['data']
+        validate_data(data)
+
+    def dynamic_validation(self, request: Request):
+        # we can not add a Context with already existent NAME and VERSION
+        # since a Context needs to be identified by seqNo
+        self._validate_request_type(request)
+        identifier, req_id, operation = get_request_data(request)
+        context_name = get_write_context_name(request)
+        context_version = get_write_context_version(request)
+        path = ContextHandler.make_state_path_for_context(identifier, context_name, context_version)
+        context, _, _ = self.get_from_state(path)
+        if context:
+            self.write_req_validator.validate(request,
+                                              [AuthActionEdit(txn_type=SET_CONTEXT,
+                                                              field='*',
+                                                              old_value='*',
+                                                              new_value='*')])
+        else:
+            self.write_req_validator.validate(request,
+                                              [AuthActionAdd(txn_type=SET_CONTEXT,
+                                                             field='*',
+                                                             value='*')])
+
+    def gen_txn_id(self, txn):
+        self._validate_txn_type(txn)
+        path = ContextHandler.prepare_context_for_state(txn, path_only=True)
+        return path.decode()
+
+    def update_state(self, txn, prev_result, request, is_committed=False) -> None:
+        self._validate_txn_type(txn)
+        path, value_bytes = ContextHandler.prepare_context_for_state(txn)
+        self.state.set(path, value_bytes)
+
+    @staticmethod
+    def make_state_path_for_context(authors_did, context_name, context_version) -> bytes:
+        return "{DID}:{MARKER}:{CONTEXT_NAME}:{CONTEXT_VERSION}" \
+            .format(DID=authors_did,
+                    MARKER=MARKER_CONTEXT,
+                    CONTEXT_NAME=context_name,
+                    CONTEXT_VERSION=context_version).encode()
+
+    @staticmethod
+    def prepare_context_for_state(txn, path_only=False):
+        origin = get_from(txn)
+        context_name = get_txn_context_name(txn)
+        context_version = get_txn_context_version(txn)
+        value = {
+            META: get_txn_context_meta(txn),
+            DATA: get_txn_context_data(txn)
+        }
+        path = ContextHandler.make_state_path_for_context(origin, context_name, context_version)
+        if path_only:
+            return path
+        seq_no = get_seq_no(txn)
+        txn_time = get_txn_time(txn)
+        value_bytes = encode_state_value(value, seq_no, txn_time)
+        return path, value_bytes
+
+
+def _bad_uri(uri_string):
+    url = findall(URI_REGEX, uri_string)
+    if not url:
+        return True
+    return False
+
+
+def validate_data(data):
```

Comment:
`validate_data` and `validate_meta` looks like Schema-based static validation, so it's better to do it there (within `ClientSetContextOperation` in `types.py`).
In general the static validation should be done as follows:
- validation that we have expected fields of expected types should be done as Schema-based validation in `types.py`
- request-specific business logic which doesn't depend on the current state and order of transactions should be done in `static_validation` in Request Handler classes.

Reply:
fixed in [51fd316](https://github.com/hyperledger/indy-node/pull/1430/commits/51fd316b9346906f925fe1cd0f08780983f9387f)